CLI: Add Yarn workspaces support for init command

pocka opened the pull request:
Issue: close #9045 

## What I did

Add support Yarn workspaces support for init command.

I changed `@storybook/cli init` command to lookup `yarn.lock` in ancestor directories, in addition to the current directory.

```
# Run `@storybook/cli init` on subdir

## use npm
.
├── package.json
├── package-lock.json
└── subdir/

.
├── package.json
├── package-lock.json
└── subdir/
    ├── package.json
    └── package-lock.json

.
├── package.json
├── yarn.lock
└── subdir/
    ├── package.json
    └── package-lock.json

## use yarn
.
├── package.json
├── yarn.lock
└── subdir/

.
├── package.json
├── package-lock.json
└── subdir/
    ├── package.json
    └── yarn.lock.json
```

I believe this commit doesn't break normal use-cases, but it surely adds complexity.
Should I implement `--use-yarn` flag instead?

## How to test

- Is this testable with Jest or Chromatic screenshots? ... no
- Does this need a new example in the kitchen sink apps? ... no
- Does this need an update to the documentation? ... Updated README

If your answer is yes to any of these, please make sure to include it in your PR.

<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to current release. Storybook maintainers cherry-pick bug and documentation fixes into the `master` branch as part of the release process, so you shouldn't need to worry about this.

Maintainers: Please tag your pull request with at least one of the following:
`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

-->

pocka committed:
CLI: Add Yarn workspaces support for init command

This commit changes how the cli decide which package manager to use.

Previously, init command sees only the current directory. But for
projects using Yarn workspaces, this detection method doesn't work,
because Yarn does not emit yarn.lock on workspace directories.

To handle this case, I modified the lookup logic to also see ancestor
directories.

https://github.com/storybookjs/storybook/issues/9045

pocka committed:
CLI: Update README

ndelangen commented on lib/cli/lib/has_yarn.js:
```diff
@@ -1,13 +1,40 @@
+ * return true if it exists.
+ */
+function hasLockFile() {
+  const f = finder();
```
we could use `find-up` here?

pocka replied:
Thanks for your advice!

pocka committed:
CLI: Use find-up for lookup yarn.lock

Use find-up instead of find-package-json because of simplicity and
performance (search lockfiles directly = less file I/O).

shilman reviewed (approved):
LGTM!!
